Zend/zend_API.c

```diff
@@ -1846,6 +1846,73 @@ ZEND_API zend_result object_init_ex(zval *arg, zend_class_entry *class_type) /*
 }
 /* }}} */
 
+ZEND_API zend_result object_init_with_constructor(zval *arg, zend_class_entry *class_type, uint32_t param_count, zval *params, HashTable *named_params) /* {{{ */
+{
+	zend_result status = _object_and_properties_init(arg, class_type, NULL);
+	if (UNEXPECTED(status == FAILURE)) {
+		ZVAL_UNDEF(arg);
+		return FAILURE;
+	}
+	zend_object *obj = Z_OBJ_P(arg);
+	zend_function *constructor = obj->handlers->get_constructor(obj);
+	if (constructor == NULL) {
+		/* The constructor can be NULL for 2 different reasons:
+		 * - It is not defined
+		 * - We are not allowed to call the constructor (e.g. private, or internal opaque class)
+		 *   and an exception has been thrown
+		 * in the former case, we are (mostly) done and the object is initialized,
+		 * in the latter we need to destroy the object as initialization failed
+		 */
+		if (UNEXPECTED(EG(exception))) {
+			zval_ptr_dtor(arg);
+			ZVAL_UNDEF(arg);
+			return FAILURE;
+		} else {
+			/* Surprisingly, this is the only case where internal classes will allow to pass extra arguments
+			 * However, if there are named arguments, an Error must be thrown to be consistent with new ClassName() */
+			if (UNEXPECTED(named_params != NULL)) {
+				/* Throw standard Error */
+				zend_string *arg_name = NULL;
+					zend_hash_get_current_key(named_params, &arg_name, /* num_index */ NULL);
+				ZEND_ASSERT(arg_name != NULL);
```

Comment:
I think it's actually allowed to pass an empty hash table to the underlying `zend_call_known_function` function.
And API users might expect the same to hold for this function.
However, in that case this line will fail.

Reply:
Indeed, fixed this.